status/condition_set.go

```diff
@@ -0,0 +1,244 @@
+	c.object.SetConditions(conditions)
+}
+
+// RemoveCondition removes the non normal condition that matches the ConditionType
+// Not implemented for normal conditions
+func (c ConditionSet) Clear(t string) error {
+	var conditions []Condition
+
+	if c.object == nil {
+		return nil
+	}
+	// Normal conditions are not handled as they can't be nil
+	if t == c.root || lo.Contains(c.dependents, t) {
+		return fmt.Errorf("clearing normal conditions not implemented")
+	}
+	cond := c.Get(t)
+	if cond == nil {
+		return nil
+	}
+	for _, c := range c.object.GetConditions() {
+		if c.Type != t {
+			conditions = append(conditions, c)
+		}
+	}
+
+	// Sorted for convenience of the consumer, i.e. kubectl.
+	sort.Slice(conditions, func(i, j int) bool { return conditions[i].Type < conditions[j].Type })
+	c.object.SetConditions(conditions)
+
+	return nil
+}
+
+// SetTrue sets the status of t to true with the reason, and then marks the root condition to
+// true if all other dependents are also true.
+func (c ConditionSet) SetTrue(conditionType string) {
+	c.SetTrueWithReason(conditionType, conditionType, conditionType)
+}
+
+// SetTrueWithReason sets the status of t to true with the reason, and then marks the root condition to
+// true if all other dependents are also true.
+func (c ConditionSet) SetTrueWithReason(conditionType string, reason, message string) {
+	c.Set(Condition{
+		Type:    conditionType,
+		Status:  metav1.ConditionTrue,
+		Reason:  reason,
+		Message: message,
+	})
+	c.recomputeRootCondition(conditionType)
+}
+
+// SetUnknown sets the status of conditionType to Unknown and also sets the root condition
+// to Unknown if no other dependent condition is in an error state.
+func (r ConditionSet) SetUnknown(conditionType string) {
+	// set the specified condition
+	r.Set(Condition{
+		Type:    conditionType,
+		Status:  metav1.ConditionUnknown,
+		Reason:  "AwaitingReconciliation",
+		Message: "object is awaiting reconciliation",
+	})
+	r.recomputeRootCondition(conditionType)
+}
+
+// recomputeRootCondition marks the root condition to true if all other dependents are also true.
+func (r ConditionSet) recomputeRootCondition(conditionType string) {
+	if condition := r.findUnhappyDependent(); condition != nil {
+		// Propagate unroot dependent to root condition.
+		r.Set(Condition{
+			Type:    r.root,
+			Status:  condition.Status,
+			Reason:  condition.Reason,
+			Message: condition.Message,
+		})
+	} else if conditionType != r.root {
+		// Set the root condition to true.
+		r.SetTrue(r.root)
+	}
+}
+
+func (c ConditionSet) findUnhappyDependent() *Condition {
+	// This only works if there are dependents.
+	if len(c.dependents) == 0 {
+		return nil
+	}
+
+	// Get dependent conditions
+	conditions := lo.Filter(c.object.GetConditions(), func(condition Condition, _ int) bool {
+		return lo.Contains(c.dependents, condition.Type)
+	})
+
+	// Sort set conditions by time.
```

Comment:
How much do we care about what the dependent is that triggered the unhappiness for the ready condition? From what I can tell, we only care _that_ we have an unhappy dependent. Which one it is doesn't seem to matter as much

Reply:
Ideally the unhappy dependent should roll up into the root condition. If there are multiple, we pick the first.

Reply:
Ah, I got it. That's the piece I missed. Then, I think that this makes sense.